spec/API_specification/set_functions.md

```diff
@@ -47,9 +47,50 @@ Each `nan` value should have a count of one, while the counts for signed zeros s
     -   a namedtuple `(values, indices, inverse_indices, counts)` whose
 
         -   first element must have the field name `values` and must be an array containing the unique elements of `x`. The array must have the same data type as `x`.
-        -   second element must have the field name `indices` and must be an array containing the indices (first occurrences) of `x` that result in `values`. The array must have the same shape as `values` and must have the default integer data type.
-        -   third element must have the field name `inverse_indices` and must be an array containing the indices of `values` that reconstruct `x`. The array must have the same shape as `x` and must have the default integer data type.
-        -   fourth element must have the field name `counts` and must be an array containing the number of times each unique element occurs in `x`. The returned array must have same shape as `values` and must have the default integer data type.
+        -   second element must have the field name `indices` and must be an array containing the indices (first occurrences) of `x` that result in `values`. The array must have the same shape as `values` and must have the default integer data type (or promoted if needed).
```

Comment:
How would promotion work? Would an implementing library need to internally copy to a new array upon overflow?

Here, would not seem that an implementing library would be able to statically know what the output dtype should be until actually processing the data. Suppose the input array has `2**33` elements (`>2**31-1`) and the default integer data type is `int32`. If all elements are unique, then counts for all would be `1`; while if all are the same, then we'd have a single count which would trigger promotion.

Instead of promotion, we could say something to the effect of "must have an `int64` dtype (i.e., the largest integer data type supported by the standard)".

Reply:
Conformant libraries are already expected to support `int64`: https://github.com/data-apis/array-api/blob/aa6c89c47c6f6f37fa0b53934b3c5784985d0628/spec/API_specification/data_types.md#data-types

This have require host emulation for 32-bit systems, but support is still expected.

If need be, we could add a disclaimer, but the spec does not currently allow leeway for conforming libraries to elide certain dtypes.

Reply:
> How would promotion work? Would an implementing library need to internally copy to a new array upon overflow?
>
> Here, would not seem that an implementing library would be able to statically know what the output dtype should be until actually processing the data. 

Yes, but most likely `unique_*` will be implemented in two passes (so splitting `unique` gives some room for optimization), so by the time we need to allocate space for indices/counts, it's statically known.

> Suppose the input array has `2**33` elements (`>2**31-1`) and the default integer data type is `int32`. If all elements are unique, then counts for all would be `1`; while if all are the same, then we'd have a single count which would trigger promotion.

Yes.

> Instead of promotion, we could say something to the effect of "must have an `int64` data type (i.e., the largest integer data type supported by the standard)".

We could say that too, I don't have strong opinions here, my point is saying "default integer type" is not sufficient 🙂

Reply:
Originally, when writing the `unique` specification, the output dtype was the "default index data type". I wonder if we need to revive that distinction. Namely, that an array library should have three default data types:

1.  floating-point data type
2.  integer data type
3.  index data type

Here, having a default index data type makes sense, as counts should align accordingly (i.e., a count should never exceed the maximum array index).

Furthermore, while it may often be the case that indices will have the same dtype as the default integer dtype, this need not be the case. For example, indices may be `int64`, while the default integer dtype could be `int32` due to better target hardware support (e.g., GPUs).

Thoughts, @rgommers?

Reply:
@kgryte I have a suggestion. How about removing all changes on the returned integer type, and moving the discussion to a new issue? I suppose the decision made here will affect all routines returning the indices of array elements satisfying a certain condition. On top of my head I know `argsort` is one of them (in fact it still says "default array index data type").

Reply:
Did more homework about "array index type" and raised an issue in #319. I believe this was an overlook during a doc refactoring.

Reply:
> How about removing all changes on the returned integer type, and moving the discussion to a new issue? This way the rest of this PR is unambiguous.

Cleaned up in f1557a1.

Reply:
Thanks, @leofang!